### PR TITLE
Store active account settings per project

### PR DESCRIPTION
**Pull request by pkukielka**
Fixes https://linear.app/sourcegraph/issue/QA-188/jetbrains-adding-password-in-keychain-will-logout-user-when-same

## Changes

There is a lot of cosmetic changes due to slightly different interfaces but main changes are there:
https://github.com/sourcegraph/jetbrains/pull/2817/files#diff-56b9b0d56648edfc3a3154ee5357aeaa891fb834b049e5eb0fb1da4a2eff5c2f

This PR changes the way account settings are stored.
Instead of having one settings shared by all cody clients we keep them separated per project.
This simplifies synchronisation between instance (as it is not needed) and is more in-line with model where each agent is separate being. I believe it is also more intuitive to the users.

## Test plan

TBD

### CODE REVIEW

File: src/main/kotlin/com/sourcegraph/cody/auth/CodyAccountService.kt [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  fun loadAccount(serverUrl: String): CodyAccount {
+    val token = serverUrl.let { getFromSecureStore(it) }
+    return CodyAccount(SourcegraphServerPath(serverUrl), token)
+  }
+
+  fun storeAccount(account: CodyAccount) {
+    writeToSecureStore(account.server.url, account.token)
+  }
+
+  fun getActiveAccount(): CodyAccount? {
```
Comment (dominiccooney) [inline src/main/kotlin/com/sourcegraph/cody/auth/CodyAccountService.kt:34]: When we do JetBrains-specific state that's not coordinated with TypeScript/agent, we usually end up regretting it.

What happens if the project active account marker points to a signed out account?

File: src/main/kotlin/com/sourcegraph/cody/config/notification/CodySettingChangeListener.kt [hunk 2]
```diff
@@ -20,7 +20,7 @@ class CodySettingChangeListener(project: Project) : ChangeListener(project) {
         object : CodySettingChangeActionNotifier {
           override fun afterAction(context: CodySettingChangeContext) {
             // Notify JCEF about the config changes
-            javaToJSBridge?.callJS("pluginSettingsChanged", ConfigUtil.getConfigAsJson())
+            javaToJSBridge?.callJS("pluginSettingsChanged", ConfigUtil.getConfigAsJson(project))
```
Comment (dominiccooney) [inline src/main/kotlin/com/sourcegraph/cody/config/notification/CodySettingChangeListener.kt:23]: Nothing listens to this, so can we simply remove it?

**Review by dominiccooney — CHANGES_REQUESTED** [review]
In general I like the direction of making the settings per-project.
- Let's remove the unused webview notification.
- Please clarify the PR description; it doesn't specify whether you're talking about runtime or on-disk, it makes it sound like you'd have to sign in once for each project. Please clarify the runtime aspect of it.
- Add a manual test plan for removing an account that a project has cached as its preferred account. Does the behavior make sense, when you reopen the project?

**Comment by dominiccooney** [comment]
I assume this was started a while ago, which is why the PR is here. But let's start new work in sourcegraph/cody.

File: src/main/kotlin/com/sourcegraph/cody/config/notification/CodySettingChangeListener.kt (hunk 2 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/cody/config/notification/CodySettingChangeListener.kt:23]: For code listening on that you would need to look at `sourcegraph/sourcegraph`.
But perhaps it is an overkill, I think popup we are using for code search is recreated every time. I will check if there is no performance hit in refactoring it a bit to not require that.

Comment (dominiccooney) [reply src/main/kotlin/com/sourcegraph/cody/config/notification/CodySettingChangeListener.kt:23]: As long as it is used, that's fine. Thanks for the pointer.

**Pull request closed without merging**